plugins/inputs/pulsar_consumer/pulsar_consumer.go

```diff
@@ -0,0 +1,248 @@
+  name = ""
+
+  ## Is consumer exclusive
+  # exclusive = false
+
+  ## Queue size for this consumer
+  # queue_size = 1000
+
+  ## Path to certificates and key for TLS
+  # tls_ca = ""
+  # tls_cert = ""
+  # tls_key = ""
+
+  ## Other optionals
+  ping_frequency = "1s"
+  ping_timeout = "1s"
+  initial_reconnect_delay = "3s"
+  max_reconnect_delay = "10s"
+  new_consumer_timeout = "10s"
+
+  ## Maximum messages to read from the broker that have not been written by an
+  ## output.  For best throughput set based on the number of metrics within
+  ## each message and the size of the output's metric_batch_size.
+  ##
+  ## For example, if each message from the queue contains 10 metrics and the
+  ## output metric_batch_size is 1000, setting this to 100 will ensure that a
+  ## full batch is collected and the write is triggered immediately without
+  ## waiting until the next flush_interval.
+  # max_undelivered_messages = 1000
+
+  ## Data format to consume.
+  ## Each data format has its own unique set of configuration options, read
+  ## more about them here:
+  ## https://github.com/influxdata/telegraf/blob/master/docs/DATA_FORMATS_INPUT.md
+  data_format = "influx"
+`
+
+func (p *pulsarConsumer) SampleConfig() string {
+	return sampleConfig
+}
+
+func (p *pulsarConsumer) Description() string {
+	return "Read metrics from Pulsar topic"
+}
+
+func (p *pulsarConsumer) SetParser(parser parsers.Parser) {
+	p.parser = parser
+}
+
+func (p *pulsarConsumer) Start(acc telegraf.Accumulator) error {
+	p.acc = acc.WithTracking(p.MaxUndeliveredMessages)
+	ctx, cancel := context.WithCancel(context.Background())
+	p.cancel = cancel
+
+	if p.connPool == nil && p.consumer == nil {
+		p.in = make(chan pulsar.Message, p.MaxUndeliveredMessages)
+		p.errs = make(chan error)
+
+		var err error
+		p.recvTimeout, err = time.ParseDuration(p.RecvTimeout)
+
+		conf := pulsar.ManagedConsumerConfig{}
+		conf.Addr = p.URL
+		conf.Errs = p.errs
+		conf.DialTimeout, err = time.ParseDuration(p.DialTimeout)
+		conf.ConnectTimeout = conf.DialTimeout
+		conf.PingFrequency, err = time.ParseDuration(p.PingFrequency)
+		conf.PingTimeout, err = time.ParseDuration(p.PingTimeout)
+		conf.InitialReconnectDelay, err = time.ParseDuration(p.InitialReconnectDelay)
+		conf.MaxReconnectDelay, err = time.ParseDuration(p.MaxReconnectDelay)
+		conf.NewConsumerTimeout, err = time.ParseDuration(p.NewConsumerTimeout)
+		if p.TLSCA != "" && p.TLSCert != "" && p.TLSKey != "" {
+			conf.TLSConfig, err = p.TLSConfig()
+		}
+		if err != nil {
+			return err
+		}
+
+		conf.Topic = p.Topic
+		conf.Name = p.Name
+		conf.Exclusive = p.Exclusive
+		conf.QueueSize = p.QueueSize
+
+		p.connPool = pulsar.NewManagedClientPool()
+		p.consumer = pulsar.NewManagedConsumer(p.connPool, conf)
+
+		go func() {
+			err = p.consumer.ReceiveAsync(ctx, p.in)
+			if err != nil {
+				p.errs <- err
+			}
+		}()
+	}
+
+	// Start the message reader
+	p.wg.Add(1)
+	go func() {
+		defer p.wg.Done()
+		go p.receiver(ctx)
```

Comment:
I'm guessing you didn't want to goroutine this (inside the goroutine), as the `wg` would be "done" immediately